**APRENDIZ:** *Dahiana Bedoya Ramirez*

**FICHA:** *1751400*

**FECHA DE INICIO DEL CUADERNO:** *15/11/2019*

### Listas

Las principales funciones de las listas son las siguientes:

* ``list.append(x)`` -- agrega el elemento x al final de la lista.
* ``list.extend(L)`` -- agrega la lista L al final de la lista.
* ``list.insert(i,x)`` -- inserta el elemento x en la posición i.
* ``list.remove(x)`` -- remueve el elemento x.
* ``list.pop([j])`` -- remueve el elemento en la posición j de la lista.
* ``list.clear()`` -- borra la lista; elimina todos los elementos de la lista.
* ``list.index(x)`` -- devuelve el indice del elemento x
* ``list.count(x)`` -- Cuenta las veces que aparece el elemento x.
* ``list.sort(key=None,reverse=False)`` -- ordena los elementos de la lista.
* ``list.copy()`` -- crea una copia de la lista.

In [1]:
a = [1, 2, 3, 4, 5] # creación de la lista
a

[1, 2, 3, 4, 5]

In [2]:
a.append(6) # agrega el elemento al final
a

[1, 2, 3, 4, 5, 6]

In [3]:
b = [7, 8, 9]  # crea una nueva lista
a.extend(b)    # pega la nueva lista al final de la anterior
a

[1, 2, 3, 4, 5, 6, 7, 8, 9]

In [4]:
a.insert(0, 'a')  # las listas pueden contener elementos con diferentes tipos de datos
a

['a', 1, 2, 3, 4, 5, 6, 7, 8, 9]

In [5]:
x = ['a', 'b', 'b', 'c', 'c', 'c', 'd', 'd', 'd', 'd'] # cuenta la cantidad de veces que aparece cada elemento
print(x.count('a'), x.count('b'), x.count('c'), x.count('d'))

1 2 3 4


In [6]:
x.remove('c')  # borra una de las veces que aparece el elemento
x

['a', 'b', 'b', 'c', 'c', 'd', 'd', 'd', 'd']

In [7]:
a.reverse()  # invierte la lista
a

[9, 8, 7, 6, 5, 4, 3, 2, 1, 'a']

In [8]:
v = [2, 4, 1, 5, 3]  
v.sort()  # ordena la lista
v

[1, 2, 3, 4, 5]

In [9]:
sorted([2, 4, 1, 5, 3])  # devuelve una copia ordenada de la lista original

[1, 2, 3, 4, 5]

In [10]:
v.pop() # remueve el último elemento de la lista (extremo derecho).
v

[1, 2, 3, 4]

In [11]:
a.pop(2) # remueve el elemento en la posición 2 de la lista
a

[9, 8, 6, 5, 4, 3, 2, 1, 'a']

In [12]:
x = list(range(10)) # convierte un objeto rango en lista
x

[0, 1, 2, 3, 4, 5, 6, 7, 8, 9]

In [13]:
y = x  # `y` apunta en memoria a la misma lista que `x`.
y

[0, 1, 2, 3, 4, 5, 6, 7, 8, 9]

In [14]:
y.pop()  # si se remueve un elemento en la lista `y`, también se remueve en la lista `x`
x

[0, 1, 2, 3, 4, 5, 6, 7, 8]

In [15]:
x = list(range(10))  
y = x.copy()         # para evitar el problema anterior se usa `copy()`
x.pop()
y

[0, 1, 2, 3, 4, 5, 6, 7, 8, 9]

### Funciones ``filter`` y ``map``

In [16]:
# Filtra los elementos mayores que 4
# MAL
a = [3, 4, 5]
b = []
for i in a:
    if i > 4:
        b.append(i)
print(b)

[5]


In [17]:
# BIEN
a = [3, 4, 5]
b = [i for i in a if i > 4]
# O:
b = filter(lambda x: x > 4, a)


In [18]:
# Sume 3 a todos los elementos de una lista
# MAL
a = [3, 4, 5]
for i in range(len(a)):
    a[i] += 3
print(a)

[6, 7, 8]


In [19]:
# BIEN
a = [3, 4, 5]
a = [i + 3 for i in a]
a

[6, 7, 8]

In [20]:
# BIEN
a = [3, 4, 5]
a = map(lambda i: i + 3, a)
list(a)

[6, 7, 8]

### Desempaquetado de listas

In [21]:
a, *rest = [1, 2, 3]    # a = 1, rest = [2, 3]
print(rest)

[2, 3]


In [22]:
a, *middle, c = [1, 2, 3, 4] # a = 1, middle = [2, 3], c = 4
print(middle)

[2, 3]


### Uso de listas como stacks

Una lista es una estructura de datos LIFO (Last In First Out): el último elemento en entrar es el primer elemento en salir. Para simular este funcionamiento se usan las funciones append y pop.

In [23]:
stack = [3, 4, 5] # se crea una lista con elementos
stack.append(6)   # se agrega el 6 al final de la lista
stack.append(7)   # se agrega el 7 al final de la lista
stack

[3, 4, 5, 6, 7]

In [24]:
stack.pop()  # al llamar a `pop` sale primero el 7 que fue el último elemento en entrar.

7

In [25]:
stack

[3, 4, 5, 6]

In [26]:
stack.pop()

6

In [27]:
stack.pop()

5

In [28]:
stack

[3, 4]

### Uso de listas como colas

Esta estructura de datos simula la cola de un supermercado. Los elementos entran por la cola y salen por la cabeza (popleft -- pop() por el inicio de la lista). Para ello, se debe importar la función ``deque`` de la librería ``collections.``

In [29]:
from collections import deque
queue = deque(["a", "b", "c"])  # se crea una cola con elementos.
queue.append("d")               # la `d` entra por el extremo derecho
queue.append("e")               # la `e` entra por el extremo derecho
queue.popleft()                 # la `a` sale por el extremo izquierdo.


'a'

In [30]:
queue.popleft()                 # la `b` sale por el extremo izquierdo

'b'

In [31]:
queue                           # imprime el contenido de la variable queue

deque(['c', 'd', 'e'])

### List comprenhensions

Este es un mecanismo para definir la composición de una lista de forma compacta. En el siguiente ejemplo se crea una lista con los cuadrados de los números del 0 al 9.

In [32]:
squares = []
for x in range(10):
    squares.append(x**2)
    
squares

[0, 1, 4, 9, 16, 25, 36, 49, 64, 81]

Un mecanismo más simple consiste en aplicar la función lambda x:x**2 a los elementos de la lista [0, ..., 9]. Para que la función anónima sea aplicada a cada elemento de la lista se usa la función map.

In [33]:
squares = list(map(lambda x: x**2, range(10)))
squares

[0, 1, 4, 9, 16, 25, 36, 49, 64, 81]

Un mecanismo mucho más simple es integrar la definición de los elementos de la lista en el ciclo for:

In [34]:
squares = [x**2 for x in range(10)]
squares

[0, 1, 4, 9, 16, 25, 36, 49, 64, 81]

En una comprenhension, se pueden anidar ciclos for tal como se muestra a continuación. Adicionalmente, se pueden agregar condiciones usando la palabra reservada if.

In [35]:
[(x, y) for x in [1,2,3] for y in [3,1,4] if x != y]

[(1, 3), (1, 4), (2, 3), (2, 1), (2, 4), (3, 1), (3, 4)]

Note que el codigo anterior es mucho más simple que el equivalente usando ciclos for anidados como se ilustra en el siguiente código.

In [36]:
combs = []
for x in [1,2,3]:
    for y in [3,1,4]:
        if x != y:
            combs.append((x, y))

combs

[(1, 3), (1, 4), (2, 3), (2, 1), (2, 4), (3, 1), (3, 4)]

Ya que este es un mecanismo para crear listas, tambien puede ser aplicado sobre elementos del tipo string. En el siguiente ejemplo, la función strip elimina los espacio en blanco al principio y al final de la cadena de caracteres.

In [37]:
# llamada de un metodo sobre cada elemento
x = ['  a a', '  a a  ', 'a a  ']
[y.strip() for y in x]  # elimina los espacios en blanco que delimitan la cadena

['a a', 'a a', 'a a']

In [38]:
['linea ' + str(i) for i in range(1, 6)]

['linea 1', 'linea 2', 'linea 3', 'linea 4', 'linea 5']

In [39]:
for x in ['linea ' + str(i) for i in range(1, 6)]: print(x)

linea 1
linea 2
linea 3
linea 4
linea 5


In [40]:
# numeros del 1 al 20 que contienen un '1'
import re
[str(x) for x in range(1,21) if re.search('1', str(x))]

['1', '10', '11', '12', '13', '14', '15', '16', '17', '18', '19']

In [41]:
# números del 1 al 20 que finanlizan con un '1'
import re
[str(x) for x in range(1,21) if re.search('1$', str(x))]

['1', '11']

In [42]:
# cantidad de números del 1 al 20 que contienen un '1'
import re
len([str(x) for x in range(1,21) if re.search('1', str(x))])

11

In [43]:
# extrae los caracteres en las posiciones 2, 3 y 4
x = ["123456790",
     "abcdefghi",
     "jklmnopqr" ]
[m[2:5] for m in x]

['345', 'cde', 'lmn']

In [44]:
# extrae los caracteres entre corchetes `[` y `]`.
x = ["-->[456]<--",
     "-->[def]<--",
     "-->[nop]<--",
    "-------->[123456]<---------"]
[m[(m.find('[')+1):m.find(']')] for m in x]

['456', 'def', 'nop', '123456']

In [45]:
# extrae la segunda palabra de cada línea
x = ["Bash is a Unix shell and command language",
     "written by Brian Fox for the ", 
     "GNU Project as a free software",
     "replacement for the Bourne shell."]
[m.split(' ')[1] for m in x]

['is', 'by', 'Project', 'for']

A continuación se presentan varios ejemplos de iteración sobre strings.

In [46]:
## iteracción sobre strings -- MAL
nums = ""
for n in range(20):
  nums += str(n)   # Lento e ineficiente
print(nums)

012345678910111213141516171819


In [47]:
## iteración sobre strings -- BIEN
nums = []
for n in range(20):
  nums.append(str(n))
print("".join(nums))  # más eficiente

012345678910111213141516171819


In [48]:
# iteración sobre strings -- MEJOR
nums = [str(n) for n in range(20)]
print("".join(nums))

012345678910111213141516171819


In [49]:
# también se pueden crear tuplas.
[(x, x**2) for x in range(6)]

[(0, 0), (1, 1), (2, 4), (3, 9), (4, 16), (5, 25)]

En el siguiente ejemplo, el primer for recorre los elementos de la lista vec mientras que el segundo ciclo for recorre las componentes de dicho elemento.

In [50]:
vec = [[1,2,3], [4,5,6], [7,8,9]]
[num for elem in vec for num in elem]

[1, 2, 3, 4, 5, 6, 7, 8, 9]

In [51]:
from math import pi  # otro ejemplo
[str(round(pi, i)) for i in range(1, 6)]

['3.1', '3.14', '3.142', '3.1416', '3.14159']

In [52]:
## se simula una matriz como una lista cuyos elementos son listas de números.
matrix = [
    [1, 2, 3, 4],
    [5, 6, 7, 8],
    [9, 10, 11, 12],
]

In [53]:
## se calcula la transpuesta 
[[row[i] for row in matrix] for i in range(4)]

[[1, 5, 9], [2, 6, 10], [3, 7, 11], [4, 8, 12]]

In [54]:
## este es el codigo equivalente tradicional.
transposed = []
for i in range(4):
    transposed.append([row[i] for row in matrix])

transposed

[[1, 5, 9], [2, 6, 10], [3, 7, 11], [4, 8, 12]]

In [55]:
# otro codigo equivalente.
transposed = []
for i in range(4):
    # the following 3 lines implement the nested listcomp
    transposed_row = []
    for row in matrix:
        transposed_row.append(row[i])
    transposed.append(transposed_row)

transposed

[[1, 5, 9], [2, 6, 10], [3, 7, 11], [4, 8, 12]]

In [56]:
# otra forma
list(zip(*matrix))

[(1, 5, 9), (2, 6, 10), (3, 7, 11), (4, 8, 12)]

### Borrado de elementos de una lista

El comando del permite borar elementos de una lista.

In [57]:
a = [-1, 1, 66.25, 333, 333, 1234.5]
del a[0]
a

[1, 66.25, 333, 333, 1234.5]

In [58]:
del a[2:4]
a

[1, 66.25, 1234.5]

In [59]:
del a[:]
a

[]

In [60]:
del a

### Revisión

**Ejercicio.**-- Algunos sistemas Unix tienen el comando tac (reverso de cat) el cual imprime el contenido de un archivo en inverso, es decir, empezando por la última línea. Implemente esta función en Python.


**Ejercicio.**-- En Unix, el comando rev imprime el contenido de un archivo invirtiendo el orden de los caracteres de cada línea. Por ejemplo, si el archivo contiene las lineas 1234 y abcd, rev imprime 4321 y dcba. Implemente esta función en Python.


A continuación se presenta como implementar la funcionalidad del comando wc de Unix.

In [61]:
# se cuentan las líneas de un archivo
len(open('out.1','r').readlines())

FileNotFoundError: [Errno 2] No such file or directory: 'out.1'

In [ ]:
# se cuentan las palabras que tiene un archivo
x = open('out.1','r').readlines()
y = ' '.join(x)
len(y.split(' '))

In [ ]:
# se cuentan los caracteres que contiene un archivo
x = open('out.1','r').readlines()
x = ''.join(x)
len(x)

Una de las principales tareas en ciencia de los datos, es la lectura de tablas de datos para su manipulación posterior. Para generar los datos de prueba se usa el cellmagic %writefile el cual escribe el conteido de la celda al archivo especificado.

In [ ]:
%%writefile out.1
Date, Year, CustomerID, Value
2013-01-12, 2013, 1, 100
2014-05-12, 2014, 1, 100
2013-02-25, 2013, 2, 200
2013-04-04, 2013, 1, 100
2013-06-21, 2013, 2, 200
2014-05-18, 2014, 1, 100
2014-06-23, 2014, 2, 200
2013-02-28, 2013, 1, 100
2013-08-02, 2013, 1, 100

Seguidamente se carga el archivo (cada línea del archivo es almacenada como un string):

In [ ]:
x = open('out.1','r').readlines()
x

Se remueven los retornos de carro ('\n') usando una list comprenhension.

In [ ]:
x = [z.replace('\n', '') for z in x]
x

Se usan los delimitadores para separar los campos.

In [ ]:
x = [z.split(',') for z in x]
x

Cada registro es almacenado como una lista.

Ahora, por ejemplo, se agregará el campo ``Month`` extrayendo el mes del campo Date. El proceso se realizará incrementalmente para facilitar su visualización.

In [ ]:
# extrae el campo Date
[z[0] for z in x[1:]] 

In [ ]:
# separa Date en sus partes
[z[0].split('-') for z in x[1:]] 

In [ ]:
# el mes ocupa la posicion 1
[z[0].split('-')[1] for z in x[1:]] # el mes

Pega el mes al final de la lista

In [ ]:
x[1:] = [z+[z[0].split('-')[1]] for z in x[1:]] 
x

Agrega el nombre del campo y los compoentes

In [ ]:
x[0].append('Month')
x

También resulta muy simple la aplicación de filtros para seleccionar solo una porción de los datos. Por ejemplo, para obtener únicamente los registros del año 2013:

In [ ]:
[z for z in x if z[1] == ' 2013']

 **Ejercicio 1.**-- Imprima un listado de valores únicos para el campo ``City`` del archivo ``files/customer.``
 
 **Ejercicio 2.**-- Agregue el campo ``subtotal`` a la tabla ``orderline`` el cual es calculado como ``Quantity`` por ``Price``.

**Ejercicio 3.**-- Agregue el campo ``ProductName`` a la tabla ``orderline``.

**Ejercicio 4.**-- Agregue el campo ``Total`` a la tabla ``order``. Este campo se calcula como la suma de ``ProductID`` * ``Price`` para cada ``OrderID`` en la tabla ``orderline``.

**Ejercicio 5.**-- Escriba una función que implemente la funcionaldad del comando ``sort`` del ``bash``.

* Recibe el nombre de un archivo.
* Lo lee del disco duro.
* Tiene un parámetros que le indica si es orden ascendente o descendente.
* Imprime el archivo ordenado en pantalla (podría ser al disco duro si se desea emular toda la funcionalidad de sort).

**Ejercicio 6.**-- Modifique el programa anterior para poder indicar el número del campo (o campos) a partir de los cuales se realizará el ordenamiento.

**Ejercicio 7.**-- Escriba una función que implemente la funcionaldad del comando ``uniq`` del ``bash``.

* Recibe el nombre de un archivo.
* Lo lee del disco duro.
* Elimina líneas consecutivas iguales.
* Imprime el resultado en la pantalla.

**Ejercicio 8.**-- Escriba una función que implemente la funcionaldad del comando ``join`` del ``bash``.

* Recibe el nombre de dos archivos.
* Suponga que la primera columna de cada archivo contiene la clave a partir de la cual se hará el ``join``.
* Imprima el resultado en pantalla.

**Ejercicio 9.**-- Escriba una función que implemente la funcionaldad del comando ``paste`` del ``bash``.

* Recibe el nombre de dos archivos.
* Imprima el resultado en pantalla.
 

### Tuplas y secuencias

Una tupla es una secuencia de elementos separados por comas.

In [ ]:
t = 12345, 54321, 'hello!'
t[0]

In [ ]:
t  # una tupla se diferencia de una lista porque se imprime entre paréntesis. 

In [ ]:
# Las tuplas pueden anidarse.
u = t, (1, 2, 3, 4, 5)
u

In [ ]:
x, y, z = t # asignación de los elementos de una tupla.

In [ ]:
x

In [ ]:
y

In [ ]:
z

### Conjuntos

Un conjunto es una estructura de datos cuyos elementos no se repiten.

In [ ]:
basket = {'apple', 'orange', 'apple', 'pear', 'orange', 'banana'}
print(basket)                      # los elmentos duplicados se removieron.

In [ ]:
'orange' in basket # `in` se usa para determinar si el elemento está en el conjunto

In [ ]:
'crabgrass' in basket

In [ ]:
a = set('abracadabra')    # el string se descompone en sus letras.
b = set('alacazam')
a                         # letras únicas en el string `a`.

A continuación se presentan las principales operaciones entre conjuntos.

In [ ]:
a - b                              # letras en a pero no en b (diferencia)

In [ ]:
a | b                              # letras en a o en b (OR)

In [ ]:
a & b                              # letras en a y en b (intersección)

In [ ]:
a ^ b                              # letras en a o en b pero no en ambos

In [ ]:
# List comprenhensions en conjuntos
a = {x for x in 'abracadabra' if x not in 'abc'} # letras que estan en `abracadabra` y no están en `abc`.
a